### PR TITLE
🐛 Fix thirds and fill layer in Safari

### LINKED ISSUE #13433: amp-story: "lower-third" grid area incorrectly positioned
**This only appears to affect beta versions of Safari.**

The "lower-third" grid area of the "thirds" template is not positioned in the bottom third of the viewport  in beta versions of Safari. (Screenshots are from desktop Safari Technology Preview Release 49 (Safari 11.2, WebKit 13606.1.3.3); it has been reported that the version of Safari that ships with the beta version of iOS exhibits the same behavior.)

Actual (beta Safari):

screen shot 2018-02-11 at 18 32 22

Expected (stable Safari):

screen shot 2018-02-11 at 18 32 13

Source code:

```html
<!doctype html>
<html amp lang="en">

<head>
  <meta charset="utf-8">
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-story" src="https://cdn.ampproject.org/v0/amp-story-0.1.js"></script>
  <link rel="canonical" href="http://foo.com/">
  <title>Hello, World</title>

  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
</head>

<body>
  <amp-story standalone>
    <amp-story-page id="page0">
      <amp-story-grid-layer template="thirds">
        <div grid-area="lower-third">
          <p>Hello, World</p>
        </div>
      </amp-story-grid-layer>
    </amp-story-page>
  </amp-story>
</body>

</html>
```

### PR DESCRIPTION
**Pull request by Enriqe**
Fixes #13433

Even though the layer has position: absolute; bottom: 0; top: 0;, Safari collapses the height as [this comment points out](https://github.com/ampproject/amphtml/issues/13433#issuecomment-422072646). Setting it explicitly prevents this.

### CODE REVIEW

**Comment by gmajoulet** [comment]
I think this fix doesn't work, since there is another `height` attribute in the same CSS block, that will override the one you added.
How can we test your fix?

Also, should we move to positioning all our "absolute" elements with
`top: 0; left: 0; height: 100%; width: 100%`? Is there anywhere else in our codebase where this bug might exist?

**Comment by Enriqe** [comment]
Tested it with a simple test case and it works well:

```
<!doctype html>
<html amp lang="en">

<head>
  <meta charset="utf-8">
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-story" src="https://cdn.ampproject.org/v0/amp-story-1.0.js"></script>
  <link rel="canonical" href="http://foo.com/">
  <title>Hello, World</title>

  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
</head>

<body>
  <amp-story standalone>
    <amp-story-page id="page0">
      <amp-story-grid-layer template="thirds">
        <div grid-area="lower-third">
          <p>Hello, World</p>
        </div>
      </amp-story-grid-layer>
    </amp-story-page>
  </amp-story>
</body>

</html>
```

Also tested with stories that were previously broken by the previous PR and everything seemed to work fine.

Was also waiting for the visual tests to run and they certainly caught something, updated the PR.

And I just filed #19195 for the latter part of your comment.

File: extensions/amp-story/1.0/amp-story.css [hunk 1]
```diff
@@ -261,6 +261,7 @@ amp-story-cta-layer {
 
 /** Grid level */
 amp-story-grid-layer {
+  box-sizing: border-box !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: Why is this needed?

File: extensions/amp-story/1.0/amp-story.css [hunk 2]
```diff
@@ -286,13 +287,12 @@ amp-story-grid-layer {
 }
 
 .i-amphtml-story-grid-template-fill > :first-child {
-  bottom: 0 !important;
-  display: block !important;
-  height: auto !important;
+  top: 0 !important;
   left: 0 !important;
+  height: 100% !important;
+  width: 100% !important;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.css]: This will be overridden by the `width: auto` below, and will, from what I understand, break the layout. Could you please upload your test cases?

File: extensions/amp-story/1.0/amp-story.css (hunk 1 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/amp-story.css]: If we don't include it, the `thirds` template won't consider the padding for the 100% height. See screenshots below:

No `box-sizing: border-box`:
image

With `box-sizing: border-box`:
image


You can also see this in [a Percy build](url): (the right one has`box-sizing: border-box`). And I think this is what we want, is that correct @newmuis?

File: extensions/amp-story/1.0/amp-story.css (hunk 2 above)
Comment (Enriqe) [reply extensions/amp-story/1.0/amp-story.css]: Woops. In this case `width: auto` and `width: 100%` were behaving the same, though.

The tests can be found under the visual diff examples:

http://localhost:8000/examples/visual-tests/amp-story/amp-story-grid-layer-template-fill.html
http://localhost:8000/examples/visual-tests/amp-story/amp-story-grid-layer-template-thirds.html
http://localhost:8000/examples/visual-tests/amp-story/amp-story-grid-layer-template-horizontal.html
http://localhost:8000/examples/visual-tests/amp-story/amp-story-grid-layer-template-vertical.html

File: extensions/amp-story/1.0/amp-story.css (hunk 1 above)
Comment (newmuis) [reply extensions/amp-story/1.0/amp-story.css]: Yes, I believe the `box-sizing` is necessary here.

File: extensions/amp-story/1.0/amp-story.css (hunk 2 above)
Comment (newmuis) [reply extensions/amp-story/1.0/amp-story.css]: I'm not actually sure we need to change `fill` layers at all, since the implementation doesn't use `grid-area`s.

Comment (Enriqe) [reply extensions/amp-story/1.0/amp-story.css]: @newmuis you were right and the bug only affects to classes using `grid-area`s. Updated the PR to only update the thirds template's height.
